### ISSUE TITLE
dmwb_writeback_: page allocation failure: order:8

### ISSUE THREAD

**Issue by disaster123** [akiradeveloper/dm-writeboost#72]
Today i saw the following traces on two out of my 8 test machines. Both with 128GB of memory where are only 20GB in use and 100GB of memory cached.

[84351.576027] dmwb_writeback_: page allocation failure: order:8, mode:0x204010
[84351.576033] CPU: 3 PID: 11044 Comm: dmwb_writeback_ Tainted: G           O   3.18.18+38-ph #1
[84351.576035] Hardware name: Supermicro X8DT3/X8DT3, BIOS 2.1     03/17/2012
[84351.576037]  0000000000000000 ffff88006b6fb9c8 ffffffff815f3246 0000000000000001
[84351.576040]  0000000000204010 ffff88006b6fba58 ffffffff81138d22 ffff88006b6fba58
[84351.576043]  ffffffff815ee446 ffff88006b6fba20 ffff88103ffdbb48 ffff88006b6fba28
[84351.576046] Call Trace:
[84351.576055]  [<ffffffff815f3246>] dump_stack+0x46/0x58
[84351.576061]  [<ffffffff81138d22>] warn_alloc_failed+0xe2/0x150
[84351.576064]  [<ffffffff815ee446>] ? __alloc_pages_direct_compact+0x66/0x19f
[84351.576067]  [<ffffffff815eec68>] __alloc_pages_slowpath+0x6e9/0x754
[84351.576071]  [<ffffffff8113c09a>] __alloc_pages_nodemask+0x22a/0x240
[84351.576075]  [<ffffffff8117d4ee>] alloc_pages_current+0x9e/0x110
[84351.576079]  [<ffffffff8118615e>] new_slab+0x2ee/0x300
[84351.576082]  [<ffffffff815f0938>] __slab_alloc+0x301/0x45d
[84351.576088]  [<ffffffffa043bda0>] ? try_alloc_writeback_ios+0xb0/0x1d0 [dm_writeboost]
[84351.576091]  [<ffffffff81187f0b>] kmem_cache_alloc+0xfb/0x130
[84351.576095]  [<ffffffffa043bd81>] ? try_alloc_writeback_ios+0x91/0x1d0 [dm_writeboost]
[84351.576099]  [<ffffffffa043bda0>] try_alloc_writeback_ios+0xb0/0x1d0 [dm_writeboost]
[84351.576103]  [<ffffffffa043de85>] writeback_daemon_proc+0xb5/0x320 [dm_writeboost]
[84351.576108]  [<ffffffff810b0bb0>] ? finish_wait+0x80/0x80
[84351.576112]  [<ffffffffa043ddd0>] ? flush_proc+0x290/0x290 [dm_writeboost]
[84351.576116]  [<ffffffff81090af9>] kthread+0xc9/0xe0
[84351.576120]  [<ffffffff81090a30>] ? kthread_worker_fn+0x150/0x150
[84351.576123]  [<ffffffff815f9008>] ret_from_fork+0x58/0x90
[84351.576126]  [<ffffffff81090a30>] ? kthread_worker_fn+0x150/0x150

**Comment by akiradeveloper** [comment]
Did you try to change nr_max_batched_writeback during the run?

**Comment by akiradeveloper** [comment]
I can't help you because you didn't report

* ***what happened to your system?*** Halt? 
* dmesg
* dmsetup status

Why your program ran into such path? ***What did you try to do?***

I guess:

* Your program frequently changes the nr_max_batched_writeback to test it.
* Now most of the memory is used
* We ran into the try_alloc_writeback_ios because nr_max_batched_writeback
* And memory allocation failed (this is benign. Please see the code)
* The system is running as usual. 

I am not a superman. I can't do anything with only a fragment of dmesg.

**Comment by akiradeveloper** [comment]
@disaster123 I've just added __GFP_NOWARN to suppress this benign warning. Could you please try fix-alloc-flag branch? I think it was a very long stress test.

**Comment by disaster123** [comment]
Yes sorry tried to change the max parameter.

Today i had one while trying to create a new device:
[ 1515.268107] dmsetup: page allocation failure: order:4, mode:0x40d0
[ 1515.268111] CPU: 1 PID: 28879 Comm: dmsetup Tainted: G           O   3.18.18+40-ph #1
[ 1515.268112] Hardware name: Supermicro X10SRH/X10SRH-CF, BIOS T20141128183646 11/28/2014
[ 1515.268113]  0000000000000000 ffff880073923778 ffffffff815f3246 0000000000000001
[ 1515.268115]  00000000000040d0 ffff880073923808 ffffffff81138d22 ffff880073923808
[ 1515.268116]  ffffffff815ee569 ffff880c7ffd8d80 ffff880c00000002 ffff880000000000
[ 1515.268117] Call Trace:
[ 1515.268123]  [<ffffffff815f3246>] dump_stack+0x46/0x58
[ 1515.268127]  [<ffffffff81138d22>] warn_alloc_failed+0xe2/0x150
[ 1515.268129]  [<ffffffff815ee569>] ? __alloc_pages_direct_compact+0x189/0x19f
[ 1515.268130]  [<ffffffff815eec68>] __alloc_pages_slowpath+0x6e9/0x754
[ 1515.268132]  [<ffffffff8113c09a>] __alloc_pages_nodemask+0x22a/0x240
[ 1515.268134]  [<ffffffff8117d4ee>] alloc_pages_current+0x9e/0x110
[ 1515.268135]  [<ffffffff81138e7e>] alloc_kmem_pages+0xe/0x10
[ 1515.268138]  [<ffffffff81153e88>] kmalloc_order+0x18/0x50
[ 1515.268139]  [<ffffffff81153ef1>] kmalloc_order_trace+0x31/0xb0
[ 1515.268142]  [<ffffffffa03367fc>] ? large_array_alloc+0xcc/0x180 [dm_writeboost]
[ 1515.268144]  [<ffffffffa03367fc>] large_array_alloc+0xcc/0x180 [dm_writeboost]
[ 1515.268146]  [<ffffffffa0337100>] resume_cache+0x240/0x1480 [dm_writeboost]
[ 1515.268147]  [<ffffffffa0336640>] ? writeboost_map+0x620/0x620 [dm_writeboost]
[ 1515.268149]  [<ffffffffa0334fe0>] writeboost_ctr+0x270/0x650 [dm_writeboost]
[ 1515.268156]  [<ffffffffa010a008>] ? dm_split_args+0x58/0x160 [dm_mod]
[ 1515.268159]  [<ffffffffa010a2a4>] dm_table_add_target+0x194/0x440 [dm_mod]
[ 1515.268162]  [<ffffffffa010d59d>] table_load+0xdd/0x350 [dm_mod]
[ 1515.268164]  [<ffffffffa010d4c0>] ? table_clear+0xe0/0xe0 [dm_mod]
[ 1515.268166]  [<ffffffffa010e46c>] ctl_ioctl+0x24c/0x4f0 [dm_mod]
[ 1515.268169]  [<ffffffff812d5962>] ? SYSC_semtimedop+0x232/0xbd0
[ 1515.268172]  [<ffffffffa010e723>] dm_ctl_ioctl+0x13/0x20 [dm_mod]
[ 1515.268175]  [<ffffffff811b3f13>] do_vfs_ioctl+0x83/0x500
[ 1515.268177]  [<ffffffff811a3d8e>] ? ____fput+0xe/0x10
[ 1515.268179]  [<ffffffff8108ee0c>] ? task_work_run+0xac/0xe0
[ 1515.268180]  [<ffffffff811b43dc>] SyS_ioctl+0x4c/0x90
[ 1515.268182]  [<ffffffff815f90b2>] system_call_fastpath+0x12/0x17
[ 1515.268183] Mem-Info:
[ 1515.268183] Node 0 DMA per-cpu:
[ 1515.268184] CPU    0: hi:    0, btch:   1 usd:   0
[ 1515.268185] CPU    1: hi:    0, btch:   1 usd:   0
[ 1515.268185] CPU    2: hi:    0, btch:   1 usd:   0
[ 1515.268186] CPU    3: hi:    0, btch:   1 usd:   0
[ 1515.268186] CPU    4: hi:    0, btch:   1 usd:   0
[ 1515.268187] CPU    5: hi:    0, btch:   1 usd:   0
[ 1515.268188] CPU    6: hi:    0, btch:   1 usd:   0
[ 1515.268188] CPU    7: hi:    0, btch:   1 usd:   0
[ 1515.268189] CPU    8: hi:    0, btch:   1 usd:   0
[ 1515.268189] CPU    9: hi:    0, btch:   1 usd:   0
[ 1515.268190] CPU   10: hi:    0, btch:   1 usd:   0
[ 1515.268190] CPU   11: hi:    0, btch:   1 usd:   0
[ 1515.268191] Node 0 DMA32 per-cpu:
[ 1515.268192] CPU    0: hi:  186, btch:  31 usd:   0
[ 1515.268192] CPU    1: hi:  186, btch:  31 usd:   0
[ 1515.268193] CPU    2: hi:  186, btch:  31 usd:   0
[ 1515.268193] CPU    3: hi:  186, btch:  31 usd:   0
[ 1515.268194] CPU    4: hi:  186, btch:  31 usd:   0
[ 1515.268194] CPU    5: hi:  186, btch:  31 usd:  72
[ 1515.268195] CPU    6: hi:  186, btch:  31 usd:   0
[ 1515.268196] CPU    7: hi:  186, btch:  31 usd:   0
[ 1515.268196] CPU    8: hi:  186, btch:  31 usd:   0
[ 1515.268197] CPU    9: hi:  186, btch:  31 usd:   0
[ 1515.268197] CPU   10: hi:  186, btch:  31 usd:   0
[ 1515.268198] CPU   11: hi:  186, btch:  31 usd:   0
[ 1515.268198] Node 0 Normal per-cpu:
[ 1515.268204] CPU    0: hi:  186, btch:  31 usd:   0
[ 1515.268205] CPU    1: hi:  186, btch:  31 usd:   0
[ 1515.268218] CPU    2: hi:  186, btch:  31 usd:   0
[ 1515.268219] CPU    3: hi:  186, btch:  31 usd:   0
[ 1515.268219] CPU    4: hi:  186, btch:  31 usd:   0
[ 1515.268220] CPU    5: hi:  186, btch:  31 usd:   1
[ 1515.268220] CPU    6: hi:  186, btch:  31 usd:   0
[ 1515.268221] CPU    7: hi:  186, btch:  31 usd:   0
[ 1515.268221] CPU    8: hi:  186, btch:  31 usd:   0
[ 1515.268222] CPU    9: hi:  186, btch:  31 usd:   0
[ 1515.268223] CPU   10: hi:  186, btch:  31 usd:   0
[ 1515.268223] CPU   11: hi:  186, btch:  31 usd:   0
[ 1515.268225] active_anon:1362430 inactive_anon:143555 isolated_anon:0
[ 1515.268225]  active_file:67329 inactive_file:8962673 isolated_file:12
[ 1515.268225]  unevictable:0 dirty:318003 writeback:0 unstable:0
[ 1515.268225]  free:94514 slab_reclaimable:304592 slab_unreclaimable:61788
[ 1515.268225]  mapped:12000 shmem:87 pagetables:5210 bounce:0
[ 1515.268225]  free_cma:0
[ 1515.268227] Node 0 DMA free:15392kB min:20kB low:24kB high:28kB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:15988kB managed:15904kB mlocked:0kB dirty:0kB writeback:0kB mapped:0kB shmem:0kB slab_reclaimable:0kB slab_unreclaimable:0kB kernel_stack:0kB pagetables:0kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:0 all_unreclaimable? yes
[ 1515.268231] lowmem_reserve[]: 0 1847 48180 48180
[ 1515.268243] Node 0 DMA32 free:187852kB min:2512kB low:3140kB high:3768kB active_anon:65704kB inactive_anon:69488kB active_file:11712kB inactive_file:1176220kB unevictable:0kB isolated(anon):0kB isolated(file):48kB present:1971404kB managed:1894244kB mlocked:0kB dirty:33352kB writeback:0kB mapped:2544kB shmem:12kB slab_reclaimable:162080kB slab_unreclaimable:34800kB kernel_stack:3200kB pagetables:432kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:284 all_unreclaimable? no
[ 1515.268249] lowmem_reserve[]: 0 0 46333 46333
[ 1515.268261] Node 0 Normal free:174812kB min:62996kB low:78744kB high:94492kB active_anon:5384016kB inactive_anon:504732kB active_file:257604kB inactive_file:34674472kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:48234496kB managed:47445800kB mlocked:0kB dirty:1238660kB writeback:0kB mapped:45456kB shmem:336kB slab_reclaimable:1056288kB slab_unreclaimable:212352kB kernel_stack:35072kB pagetables:20408kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:544 all_unreclaimable? no
[ 1515.268266] lowmem_reserve[]: 0 0 0 0
[ 1515.268267] Node 0 DMA: 0*4kB 0*8kB 0*16kB 1*32kB (U) 2*64kB (U) 1*128kB (U) 1*256kB (U) 1*512kB (U) 0*1024kB 1*2048kB (R) 3*4096kB (M) = 15392kB
[ 1515.268273] Node 0 DMA32: 921*4kB (UEM) 668*8kB (UEMR) 429*16kB (UEMR) 318*32kB (UEMR) 717*64kB (UEMR) 115*128kB (UMR) 390*256kB (MR) 1*512kB (R) 1*1024kB (R) 0*2048kB 0*4096kB = 188052kB
[ 1515.268278] Node 0 Normal: 19539*4kB (UEM) 3101*8kB (UEM) 1783*16kB (UEM) 1244*32kB (UEM) 64*64kB (UM) 3*128kB (UM) 0*256kB 0*512kB 1*1024kB (R) 0*2048kB 0*4096kB = 176804kB
[ 1515.268284] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[ 1515.268284] 9029956 total pagecache pages
[ 1515.268285] 0 pages in swap cache
[ 1515.268286] Swap cache stats: add 0, delete 0, find 0/0
[ 1515.268287] Free swap  = 3905532kB
[ 1515.268287] Total swap = 3905532kB
[ 1515.268288] 12555472 pages RAM
[ 1515.268288] 0 pages HighMem/MovableOnly
[ 1515.268289] 216485 pages reserved
[ 1515.268289] 0 pages hwpoisoned
[ 1515.268290] device-mapper: writeboost: Failed to allocate part->memory
[ 1515.270228] ------------[ cut here ]------------
[ 1515.271538] kernel BUG at mm/slub.c:3334!
[ 1515.272697] invalid opcode: 0000 [#1] SMP 
[ 1515.273973] Modules linked in: dm_writeboost(O) netconsole ipt_REJECT nf_reject_ipv4 xt_multiport cpufreq_userspace iptable_filter cpufreq_powersave ip_tables cpufreq_stats x_tables cpufreq_conservative bonding ext2 coretemp 8021q garp fuse sb_edac edac_core i2c_i801 xhci_pci xhci_hcd shpchp ipmi_si ipmi_msghandler button dm_mod raid1 md_mod usbhid usb_storage ohci_hcd bcache sg sd_mod ehci_pci ehci_hcd usbcore usb_common ahci libahci igb ixgbe(O) mpt3sas i2c_algo_bit vxlan raid_class ip6_udp_tunnel i2c_core udp_tunnel scsi_transport_sas ptp pps_core
[ 1515.290819] CPU: 1 PID: 28879 Comm: dmsetup Tainted: G           O   3.18.18+40-ph #1
[ 1515.293054] Hardware name: Supermicro X10SRH/X10SRH-CF, BIOS T20141128183646 11/28/2014
[ 1515.295355] task: ffff880934639d30 ti: ffff880073920000 task.ti: ffff880073920000
[ 1515.297492] RIP: 0010:[<ffffffff811878a3>]  [<ffffffff811878a3>] kfree+0x133/0x140
[ 1515.299737] RSP: 0018:ffff880073923a68  EFLAGS: 00010246
[ 1515.301271] RAX: 002fffff80000000 RBX: ffff8803adbc0000 RCX: ffffea001ae5bc20
[ 1515.303297] RDX: 000077ff80000000 RSI: ffff880c7ffd8fc8 RDI: ffff8803adbc0000
[ 1515.305342] RBP: ffff880073923a88 R08: 0000000000000004 R09: 0000000000000001
[ 1515.307375] R10: ffffea000eb6f000 R11: 0000000000000120 R12: ffff880c2cb1b040
[ 1515.309421] R13: ffffffffa0336864 R14: ffff880c394d90e8 R15: ffff880073923b48
[ 1515.311480] FS:  00007f90e748f7a0(0000) GS:ffff880c7fc20000(0000) knlGS:0000000000000000
[ 1515.313803] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 1515.315435] CR2: ffffffffff600400 CR3: 0000000a394e4000 CR4: 00000000001407e0
[ 1515.317492] Stack:
[ 1515.318071]  ffff880934639d30 00000000000019e8 ffff880c2cb1b040 0000000000000101
[ 1515.392457]  ffff880073923ab8 ffffffffa0336864 0000000000000007 ffff880c394d9000
[ 1515.468213]  0000000000000008 0000000000000240 ffff880073923ba8 ffffffffa0337100
[ 1515.543488] Call Trace:
[ 1515.616899]  [<ffffffffa0336864>] large_array_alloc+0x134/0x180 [dm_writeboost]
[ 1515.691786]  [<ffffffffa0337100>] resume_cache+0x240/0x1480 [dm_writeboost]
[ 1515.766408]  [<ffffffffa0336640>] ? writeboost_map+0x620/0x620 [dm_writeboost]
[ 1515.841549]  [<ffffffffa0334fe0>] writeboost_ctr+0x270/0x650 [dm_writeboost]
[ 1515.917403]  [<ffffffffa010a008>] ? dm_split_args+0x58/0x160 [dm_mod]
[ 1515.993503]  [<ffffffffa010a2a4>] dm_table_add_target+0x194/0x440 [dm_mod]
[ 1516.068792]  [<ffffffffa010d59d>] table_load+0xdd/0x350 [dm_mod]
[ 1516.143177]  [<ffffffffa010d4c0>] ? table_clear+0xe0/0xe0 [dm_mod]
[ 1516.216420]  [<ffffffffa010e46c>] ctl_ioctl+0x24c/0x4f0 [dm_mod]
[ 1516.288193]  [<ffffffff812d5962>] ? SYSC_semtimedop+0x232/0xbd0
[ 1516.358476]  [<ffffffffa010e723>] dm_ctl_ioctl+0x13/0x20 [dm_mod]
[ 1516.429695]  [<ffffffff811b3f13>] do_vfs_ioctl+0x83/0x500
[ 1516.501216]  [<ffffffff811a3d8e>] ? ____fput+0xe/0x10
[ 1516.570831]  [<ffffffff8108ee0c>] ? task_work_run+0xac/0xe0
[ 1516.641244]  [<ffffffff811b43dc>] SyS_ioctl+0x4c/0x90
[ 1516.709848]  [<ffffffff815f90b2>] system_call_fastpath+0x12/0x17
[ 1516.779413] Code: 49 8b 02 31 f6 f6 c4 40 74 04 41 8b 72 68 4c 89 d7 e8 b2 31 fb ff eb 90 49 8b 42 30 49 8b 12 80 e6 80 4c 0f 45 d0 e9 33 ff ff ff <0f> 0b 66 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 48 83 fe 
[ 1516.930126] RIP  [<ffffffff811878a3>] kfree+0x133/0x140
[ 1517.003001]  RSP <ffff880073923a68>
[ 1517.076328] ---[ end trace 552a4f8da8114d09 ]---

**Comment by akiradeveloper** [comment]
What code did you use?

Do you use the same SSD? (what size?)

Try again?

**Comment by akiradeveloper** [comment]
It says it just failed to allocate memory.

So please restart and retry.

**Comment by disaster123** [comment]
Tried it multiple times. Just started working after a reboot. Also tried drop_caches - SSD 75GB. 
Code from writeboost from @onlyjob 
dmsetup create "$M" --table "0 $B writeboost $H $S"

**Comment by akiradeveloper** [comment]
What branch did you use?

**Comment by akiradeveloper** [comment]
I see no such serious degradation on my test machine.

*** What change did you do between before and after the bug is seen ***

**Comment by akiradeveloper** [comment]
One thing I suspect is you flip the HDD and SSD. Since the HDD is too big to be a caching device, so it couldn't allocate memory for the huge metadata.

Did you change the initialization script before and after the bug appears?

I want all information otherwise I can't help you.

Btw, this is too off-topic. Why didn't you make a different thread...

**Comment by disaster123** [comment]
No this was scripted using partlabels the disks are not flipped. No i did not change anything.

I was in the process of recreating 8 writeboost devices in a row. 7 succeeded the 8th was failing with a segfault in dmsetup showing the stack trace above.

**Comment by akiradeveloper** [comment]
You use 2.0.4? Please answer this question.

It was successful before so it is natural to presume that something changed.

**Comment by akiradeveloper** [comment]
You said you try to initialize devices in parallel. Please stop it for now and forever if you agree with my opinion below.

I am not sure `dmsetup create` is designed so it can be called in parallel. IMHO, it's overkill to consider locking during initialization. So I don't recommend you to do that unless you are 100% confident. I believe dm-writeboost doesn't share anything between devices but I am not fully confident.

The initialization time is now too short to care about. So, parallelism doesn't actually help any users.

**Comment by akiradeveloper** [comment]
Until I am informed of which version or branch you are using, I won't tell you anything

**Comment by disaster123** [comment]
Version / Branch 2.0.4

No i do not initialize in parallel in this cace. I do this at boot time but this one is running since days and i create one dm-wb device after another.

Currently i have 5 drives running / under load.

If i try to create the 6 drive dmsetup segfault due to a page allocation failure in the writeboost module:
[13430.827838] dmsetup: page allocation failure: order:4, mode:0x40d0
[13430.827842] CPU: 8 PID: 10848 Comm: dmsetup Tainted: G           O   3.18.18+40-ph #1
[13430.827843] Hardware name: Supermicro X10SRH/X10SRH-CF, BIOS T20141128183646 11/28/2014
[13430.827844]  0000000000000000 ffff8801c9bdf778 ffffffff815f3246 0000000000000001
[13430.827846]  00000000000040d0 ffff8801c9bdf808 ffffffff81138d22 ffff8801c9bdf808
[13430.827847]  ffffffff815ee569 ffff880c7ffd8d80 ffff880c00000002 ffff880100000000
[13430.827848] Call Trace:
[13430.827854]  [<ffffffff815f3246>] dump_stack+0x46/0x58
[13430.827858]  [<ffffffff81138d22>] warn_alloc_failed+0xe2/0x150
[13430.827860]  [<ffffffff815ee569>] ? __alloc_pages_direct_compact+0x189/0x19f
[13430.827861]  [<ffffffff815eec68>] __alloc_pages_slowpath+0x6e9/0x754
[13430.827863]  [<ffffffff8113c09a>] __alloc_pages_nodemask+0x22a/0x240
[13430.827865]  [<ffffffff8117d4ee>] alloc_pages_current+0x9e/0x110
[13430.827867]  [<ffffffff81138e7e>] alloc_kmem_pages+0xe/0x10
[13430.827870]  [<ffffffff81153e88>] kmalloc_order+0x18/0x50
[13430.827871]  [<ffffffff81153ef1>] kmalloc_order_trace+0x31/0xb0
[13430.827875]  [<ffffffffa028f7fc>] ? large_array_alloc+0xcc/0x180 [dm_writeboost]
[13430.827877]  [<ffffffffa028f7fc>] large_array_alloc+0xcc/0x180 [dm_writeboost]
[13430.827879]  [<ffffffffa0290100>] resume_cache+0x240/0x1480 [dm_writeboost]
[13430.827881]  [<ffffffffa028f640>] ? writeboost_map+0x620/0x620 [dm_writeboost]
[13430.827883]  [<ffffffffa028dfe0>] writeboost_ctr+0x270/0x650 [dm_writeboost]
[13430.827889]  [<ffffffffa018b008>] ? dm_split_args+0x58/0x160 [dm_mod]
[13430.827891]  [<ffffffffa018b2a4>] dm_table_add_target+0x194/0x440 [dm_mod]
[13430.827894]  [<ffffffffa018e59d>] table_load+0xdd/0x350 [dm_mod]
[13430.827897]  [<ffffffffa018e4c0>] ? table_clear+0xe0/0xe0 [dm_mod]
[13430.827899]  [<ffffffffa018f46c>] ctl_ioctl+0x24c/0x4f0 [dm_mod]
[13430.827902]  [<ffffffff812d5962>] ? SYSC_semtimedop+0x232/0xbd0
[13430.827905]  [<ffffffffa018f723>] dm_ctl_ioctl+0x13/0x20 [dm_mod]
[13430.827907]  [<ffffffff811b3f13>] do_vfs_ioctl+0x83/0x500
[13430.827909]  [<ffffffff811a3d8e>] ? ____fput+0xe/0x10
[13430.827911]  [<ffffffff8108ee0c>] ? task_work_run+0xac/0xe0
[13430.827912]  [<ffffffff811b43dc>] SyS_ioctl+0x4c/0x90
[13430.827914]  [<ffffffff815f90b2>] system_call_fastpath+0x12/0x17

**Comment by akiradeveloper** [comment]
OK.

Question: So the problem appears *all of the sudden* although you changed *nothing* at all?

If so this is weird (and very funny) and I bet it's not a kernel issue. I think you changed *something*.

Please answer all of my questions. Otherwise, I won't help you any further. We are living in different timezone so please note turn around time in our communication is very long. (I am living in Japan)

Questions:
* Are you really sure only one out of all devices results in failures (I suspect, all devices actually fail). You can use `dmsetup ls` to show what devices are successfully created.
* Did you change kernel configuration?
* Do you use Dmitry's management tool *without any change*? (I am not sure how you initialize device. Because you didn't share it. I reviewed Dmitry's script and I only trust it)
* Did you try with only one device to create?
* Did you try without your script but manually?
* Did you try with only one CPU enabled?

Again, if the kernel isn't change the problem is likely on your side.

**Comment by akiradeveloper** [comment]
@disaster123  Oh, I notice that you are **lying** to me.

I found that the kernel is changed:

It was "G O 3.18.18+37-ph" in #63 but  "G O 3.18.18+40-ph" in this thread.

We need to fight with the same problem hand in hand. I can't help liar.

**Comment by akiradeveloper** [comment]
I add another questions:

* Did you try to reproduce with other machines? (It's ok if it's VM)

**Comment by disaster123** [comment]
Sorry but i don't lie. It's not easy to know what's important. In this case the changes in the kernel are only in the driver/md/bcache driver so nothing related to writeboost.

Isn't the problem that large_array_alloc may fail under memory pressure and that this isn't catched?

Regarding your questions:
Are you really sure only one out of all devices results in failures (I suspect, all devices actually fail).

Yes. Script is in pseudocode:
```
set -e
for i in 1 .. 8; do
   echo "wb$i /dev/disk/by-partlabel/hdd$i /dev/disk/by-partlabel/ssd$i  writeback_threshold=70,nr_max_batched_writeback=4" >>/etc/writeboosttab
  writeboost
  mkfs.xfs /dev/mapper/wb$i
  mount /dev/mapper/wb$i /mnt/wb$i
  start ceph on /mnt/wb$i/
end
´´´
You can use dmsetup ls to show what devices are successfully created.

Those are identical - same SSD size same HDD size same parameters. The problem is "only" memory pressure after creating 4-6 devices and starting the ceph-osd. The 5th oder 6th drive mapping will fail. If you do the same on a fresh boot and not starting ceph it works.
# dmsetup ls
wb69    (253:4)
wb71    (253:0)
wb68    (253:5)
wb70    (253:1)
wb67    (253:2)
wb66    (253:3)

Did you change kernel configuration?

No - just compiled a new kernel with some bcache fixes  - as this is a production system.

Do you use Dmitry's management tool without any change? (I am not sure how you initialize device. Because you didn't share it. I reviewed Dmitry's script and I only trust it)

I added the possibility to map in parallel but in the above scripted case it still maps only one at a time. Thanks for reminding me to post the changes.

@onlyonejob
sbin/wirteboost
https://gist.github.com/disaster123/858a5964c2bbbd363bba
writeboost.init
https://gist.github.com/disaster123/7e3c44aa8f29ef77eaa2

Did you try with only one device to create?

Yes this works always fine as i said before it's sum of devices.

Did you try without your script but manually?

No but if you see the above there is really nothing special.

Did you try with only one CPU enabled?

No.

Did you try to reproduce with other machines? (It's ok if it's VM)

Yes two different servers. One with Single E5 Xeon 48GB RAM other one with Dual E5 Xeon 128GB RAM

**Comment by disaster123** [comment]
There were a similiar situation with memory allocation in the vhost net driver back in 2014 which was fixed by the following commit in the kernel:

https://kernel.googlesource.com/pub/scm/linux/kernel/git/stable/linux-stable/+/23cc5a991c7a9fb7e6d6550e65cee4f4173111c5

**Comment by disaster123** [comment]
I think the same problem happens here:
[41923.154238] dmwb_writeback_: page allocation failure: order:8, mode:0x204010
[41923.154241] CPU: 2 PID: 2674 Comm: dmwb_writeback_ Tainted: G           O   3.18.18+40-ph #1
[41923.154242] Hardware name: Supermicro X10SRH/X10SRH-CF, BIOS T20141128183646 11/28/2014
[41923.154243]  0000000000000000 ffff880aea4239c8 ffffffff815f3246 0000000000000001
[41923.154244]  0000000000204010 ffff880aea423a58 ffffffff81138d22 ffff880aea423a58
[41923.154246]  ffffffff815ee446 ffff880aea423a20 ffff880c7ffd9b38 ffff880aea423a28
[41923.154247] Call Trace:
[41923.154253]  [<ffffffff815f3246>] dump_stack+0x46/0x58
[41923.154257]  [<ffffffff81138d22>] warn_alloc_failed+0xe2/0x150
[41923.154259]  [<ffffffff815ee446>] ? __alloc_pages_direct_compact+0x66/0x19f
[41923.154260]  [<ffffffff815eec68>] __alloc_pages_slowpath+0x6e9/0x754
[41923.154262]  [<ffffffff8113c09a>] __alloc_pages_nodemask+0x22a/0x240
[41923.154264]  [<ffffffff8117d4ee>] alloc_pages_current+0x9e/0x110
[41923.154266]  [<ffffffff8118615e>] new_slab+0x2ee/0x300
[41923.154268]  [<ffffffff815f0938>] __slab_alloc+0x301/0x45d
[41923.154272]  [<ffffffffa02a5da0>] ? try_alloc_writeback_ios+0xb0/0x1d0 [dm_writeboost]
[41923.154274]  [<ffffffff81187f0b>] kmem_cache_alloc+0xfb/0x130
[41923.154276]  [<ffffffffa02a5d81>] ? try_alloc_writeback_ios+0x91/0x1d0 [dm_writeboost]
[41923.154278]  [<ffffffffa02a5da0>] try_alloc_writeback_ios+0xb0/0x1d0 [dm_writeboost]
[41923.154280]  [<ffffffffa02a7e85>] writeback_daemon_proc+0xb5/0x320 [dm_writeboost]
[41923.154283]  [<ffffffff810b0bb0>] ? finish_wait+0x80/0x80
[41923.154285]  [<ffffffffa02a7dd0>] ? flush_proc+0x290/0x290 [dm_writeboost]
[41923.154288]  [<ffffffff81090af9>] kthread+0xc9/0xe0
[41923.154289]  [<ffffffff81090a30>] ? kthread_worker_fn+0x150/0x150
[41923.154292]  [<ffffffff815f9008>] ret_from_fork+0x58/0x90
[41923.154293]  [<ffffffff81090a30>] ? kthread_worker_fn+0x150/0x150
[41923.154294] Mem-Info:
[41923.154294] Node 0 DMA per-cpu:
[41923.154295] CPU    0: hi:    0, btch:   1 usd:   0
[41923.154296] CPU    1: hi:    0, btch:   1 usd:   0
[41923.154297] CPU    2: hi:    0, btch:   1 usd:   0
[41923.154297] CPU    3: hi:    0, btch:   1 usd:   0
[41923.154298] CPU    4: hi:    0, btch:   1 usd:   0
[41923.154298] CPU    5: hi:    0, btch:   1 usd:   0
[41923.154299] CPU    6: hi:    0, btch:   1 usd:   0
[41923.154300] CPU    7: hi:    0, btch:   1 usd:   0
[41923.154300] CPU    8: hi:    0, btch:   1 usd:   0
[41923.154301] CPU    9: hi:    0, btch:   1 usd:   0
[41923.154301] CPU   10: hi:    0, btch:   1 usd:   0
[41923.154302] CPU   11: hi:    0, btch:   1 usd:   0
[41923.154302] Node 0 DMA32 per-cpu:
[41923.154303] CPU    0: hi:  186, btch:  31 usd:   0
[41923.154304] CPU    1: hi:  186, btch:  31 usd:   0
[41923.154304] CPU    2: hi:  186, btch:  31 usd:   0
[41923.154305] CPU    3: hi:  186, btch:  31 usd:   0
[41923.154306] CPU    4: hi:  186, btch:  31 usd:   0
[41923.154306] CPU    5: hi:  186, btch:  31 usd:   0
[41923.154307] CPU    6: hi:  186, btch:  31 usd:  99
[41923.154307] CPU    7: hi:  186, btch:  31 usd:   0
[41923.154308] CPU    8: hi:  186, btch:  31 usd:   0
[41923.154309] CPU    9: hi:  186, btch:  31 usd:   0
[41923.154309] CPU   10: hi:  186, btch:  31 usd:   0
[41923.154310] CPU   11: hi:  186, btch:  31 usd:   0
[41923.154310] Node 0 Normal per-cpu:
[41923.154311] CPU    0: hi:  186, btch:  31 usd:   0
[41923.154312] CPU    1: hi:  186, btch:  31 usd:   0
[41923.154312] CPU    2: hi:  186, btch:  31 usd:   0
[41923.154313] CPU    3: hi:  186, btch:  31 usd:   0
[41923.154313] CPU    4: hi:  186, btch:  31 usd:   0
[41923.154314] CPU    5: hi:  186, btch:  31 usd:   0
[41923.154315] CPU    6: hi:  186, btch:  31 usd:   0
[41923.154315] CPU    7: hi:  186, btch:  31 usd:   0
[41923.154316] CPU    8: hi:  186, btch:  31 usd:   0
[41923.154316] CPU    9: hi:  186, btch:  31 usd:   0
[41923.154317] CPU   10: hi:  186, btch:  31 usd:   0
[41923.154317] CPU   11: hi:  186, btch:  31 usd:   0
[41923.154319] active_anon:1472807 inactive_anon:328748 isolated_anon:0
[41923.154319]  active_file:4176225 inactive_file:4176303 isolated_file:0
[41923.154319]  unevictable:0 dirty:11592 writeback:0 unstable:0
[41923.154319]  free:96522 slab_reclaimable:331674 slab_unreclaimable:112063
[41923.154319]  mapped:25890 shmem:124 pagetables:8542 bounce:0
[41923.154319]  free_cma:0
[41923.154321] Node 0 DMA free:15392kB min:20kB low:24kB high:28kB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:15988kB managed:15904kB mlocked:0kB dirty:0kB writeback:0kB mapped:0kB shmem:0kB slab_reclaimable:0kB slab_unreclaimable:0kB kernel_stack:0kB pagetables:0kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:0 all_unreclaimable? yes
[41923.154324] lowmem_reserve[]: 0 1847 48180 48180
[41923.154330] Node 0 DMA32 free:188604kB min:2512kB low:3140kB high:3768kB active_anon:548456kB inactive_anon:521376kB active_file:13748kB inactive_file:13700kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:1971404kB managed:1894244kB mlocked:0kB dirty:408kB writeback:0kB mapped:536kB shmem:12kB slab_reclaimable:219104kB slab_unreclaimable:67532kB kernel_stack:25680kB pagetables:1496kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:640 all_unreclaimable? no
[41923.154333] lowmem_reserve[]: 0 0 46333 46333
[41923.154334] Node 0 Normal free:182092kB min:62996kB low:78744kB high:94492kB active_anon:5342772kB inactive_anon:793616kB active_file:16691152kB inactive_file:16691456kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:48234496kB managed:47445800kB mlocked:0kB dirty:45960kB writeback:0kB mapped:103024kB shmem:484kB slab_reclaimable:1107592kB slab_unreclaimable:380720kB kernel_stack:96608kB pagetables:32672kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:0 all_unreclaimable? no
[41923.154338] lowmem_reserve[]: 0 0 0 0
[41923.154340] Node 0 DMA: 0*4kB 0*8kB 0*16kB 1*32kB (U) 2*64kB (U) 1*128kB (U) 1*256kB (U) 1*512kB (U) 0*1024kB 1*2048kB (R) 3*4096kB (M) = 15392kB
[41923.154372] Node 0 DMA32: 12057*4kB (UEM) 3825*8kB (UEM) 141*16kB (UM) 202*32kB (UEM) 116*64kB (UM) 119*128kB (UM) 88*256kB (UEM) 87*512kB (UM) 11*1024kB (M) 0*2048kB 0*4096kB = 188540kB
[41923.154378] Node 0 Normal: 37828*4kB (UEM) 3218*8kB (UM) 141*16kB (UEM) 6*32kB (UM) 1*64kB (M) 1*128kB (M) 0*256kB 0*512kB 0*1024kB 2*2048kB (M) 0*4096kB = 183792kB
[41923.154383] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[41923.154384] 8352554 total pagecache pages
[41923.154385] 12 pages in swap cache
[41923.154386] Swap cache stats: add 2651, delete 2639, find 101/104
[41923.154386] Free swap  = 3894992kB
[41923.154387] Total swap = 3905532kB
[41923.154387] 12555472 pages RAM
[41923.154388] 0 pages HighMem/MovableOnly
[41923.154388] 216485 pages reserved
[41923.154389] 0 pages hwpoisoned
[41923.154390] SLUB: Unable to allocate memory on node -1 (gfp=0x10)
[41923.154391]   cache: dmwb_rambuf, object size: 524288, buffer size: 1048576, default order: 8, min order: 8
[41923.154392]   dmwb_rambuf debugging increased min order, use slub_debug=O to disable.
[41923.154393]   node 0: slabs: 21, objs: 21, free: 0
[41923.154399] device-mapper: writeboost: Failed to allocate writeback_segs

**Comment by akiradeveloper** [comment]
> I think the same problem happens here:
[41923.154238] dmwb_writeback_: page allocation failure: order:8, mode:0x204010
[41923.154241] CPU: 2 PID: 2674 Comm: dmwb_writeback_ Tainted: G O 3.18.18+40-ph #1
[41923.154242] Hardware name: Supermicro X10SRH/X10SRH-CF, BIOS T20141128183646 11/28/2014

How do you know this appears the same problem? And how did you get this trace? The timestamp is 40000sec but your system failed in 1500sec.

I don't think so and it must be fixed in fix-alloc-flag.

> https://kernel.googlesource.com/pub/scm/linux/kernel/git/stable/linux-stable/+/23cc5a991c7a9fb7e6d6550e65cee4f4173111c5

Very bad hack. Fallback to vmalloc is really disliked by mm maintainers. We shouldn't use vmalloc.

----

Let me once tidy up the information:

* You tested with only two devices and it was OK.
* But once you try with 8 devices. The last device failed in memory allocation.
* You think that's because of memory pressure. (Why? Where is the evidence? I want to discuss based only on evidence. Otherwise, just a waste of time...)

If your SSD is 72GB * 8, it will eat only few gigabyte for the metadata. 

New Questions for you:

> If i try to create the 6 drive dmsetup segfault due to a page allocation failure in the writeboost module: 

Why 6? (typo of 6th?) You said 8. Don't bewilder me. If the memory pressure is the cause, all 6 devices should succeed. Which is correct? We are fixing problem so don't change parameters. Really frustrating.

> I was in the process of recreating 8 writeboost devices in a row. 7 succeeded the 8th was failing with a segfault in dmsetup showing the stack trace above.

You said 7 devices succeeded? Which is correct? I don't really see what you are doing. 

> Yes two different servers. One with Single E5 Xeon 48GB RAM other one with Dual E5 Xeon 128GB RAM

Really? Both failed with 8 devices?

> wb69 (253:4)
wb71 (253:0)
wb68 (253:5)
wb70 (253:1)
wb67 (253:2)
wb66 (253:3)

I see 6 devices... but you said 6th device failed... 


> The 5th oder 6th drive mapping will fail. If you do the same on a fresh boot and not starting ceph it works.

typo. I don't get what you mean.

So it succeeded if ceph isn't activated. Then why do you think this is kernel issue??? I think Ceph must hold some structures in memory after each initialization and all you need to do it to drop them between creating devices.

**Comment by akiradeveloper** [comment]
Please clarify what you are doing and only discuss based on evidence. I am really amazed that you claim memory pressure without showing /proc/meminfo etc. So we can't do any further discussion. I can't help you. Also, you need to investigate how much memory ceph-osd grabs.

For me, writeboost fails initialization because of allocation failure is pretty natural, not a bug.

**Comment by disaster123** [comment]
I'm sorry. I never wanted to confuse you. I'm also sorry I amazed you. I'll try to clarify tonight.

Stefan

Excuse my typo sent from my mobile phone.

**Comment by akiradeveloper** [comment]
Yes, please.

I think you need to first record /proc/meminfo between each initialization.

```
set -e
for i in 1 .. 8; do
   cat /proc/meminfo >> $LOG
   ...
end
```

Also, I want to know if you can initialize all 8 osds if you don't use dm-writeboost.

**Comment by disaster123** [comment]
Just one question does meminfo help? There is enough free men in general. I think it fails cause kmalloc uses mem from slab cache and want physical contiguous/unfragmented  memory. Does meminfo show those informations?

The osds do not directly matter. Due to their heavy work in the fs you just get higher pressure on slab memory. I could also use bonnie++. Also the initialization of ceph osd always works it's dmsetup which crashes while doing the mapping.

I rechecked the kernel source and there are many places where the vmalloc fallback is used.

Stefan

Excuse my typo sent from my mobile phone.

**Comment by akiradeveloper** [comment]
We need data to share evidence, not to prove one's assumption is right.

But yes, you makes sense. large_array is quite a bad structure because it allocates too many huge chunks. So shall we first try using vmalloc for this structure?

I've just pushed a new branch fix-large-array-to-vmalloc. Please try with this one.

**Comment by disaster123** [comment]
Thanks a lot - will try it today or tomorrow. I'm just wondering why you suddenly followed my argumentation?

Isn't it like this:
1.) a storage driver should be reliable
2.) a storage driver should be fast
3.) all kmalloc calls which are expected to fail (in the code) should stay as they are but you should pass GPF_NOWARN
4.) all kmalloc calls which should not fail should pass GPF_NOWARN & GPF_NORETRY and fall back to vmalloc - to free you can use kvfree which handles the situation automatically

But reliability is more important than speed.

**Comment by akiradeveloper** [comment]
> Thanks a lot - will try it today or tomorrow.

I wish it works.

> But reliability is more important than speed.

Yes. I don't like to respond to trivial issue reporting so I want it stable first. But portability too.

> I'm just wondering why you suddenly followed my argumentation?

I am not sure what you are saying. What's does the list of sentences mean? I am just modifying codes so I think it should be, not following someone's arguments. If you like to fix dm-writeboost, you are free to send me PRs.

**Comment by akiradeveloper** [comment]
If fix-large-array-to-vmalloc solves, Stefan's problem. I will fold it into v2.0.5.

Other part of the memory allocations such as allocating rambuf and writeback_ios is enough huge to replace with vmalloc but remains as-is for now.

**Comment by onlyjob** [comment]
I was trying Stefan's patch for parallel initialisation in "writeboost" with dm-writeboost/2.0.3 (I used 2.0.3 deliberately because it still does "slow" initialisation) and I've easily reproduced "Cannot allocate memory" situation on Debian kernel `3.16.7-ckt11-1 x86_64`. On 5 _dm-wb_ devices parrallel initialisation produced tainted error in `dbsetup` on two devices that were initialised first:
~~~~
[<ffffffff8150b405>] ? dump_stack+0x41/0x51                                                                                            
 [<ffffffff811426ef>] ? warn_alloc_failed+0xdf/0x130
[...]
~~~~
Such thing never happened in sequential initialisation.

I think parallel deconstruction of dm-writeboost devices is a good idea but parallel initialisation is apparently dangerous...

**Comment by akiradeveloper** [comment]
Hi Dmitry,

I am nacking to the idea of parallel initialization as I explained.

Also, nacking to entirely parallelize deconstruction.

In your script, deconstruction has two steps:

1. write back
2. dmsetup remove

The step 1 can be safely parallelized. But don't do this for the step 2. It touches the device-mapper's common structures so I am sure if it's safe.

Step 2 is much quicker than step 1 so it's meaningless to parallelize step 2.

**Comment by disaster123** [comment]
fix-large-array-to-vmalloc  branch works fine. Also for parallel initialisation of 10 devices.

**Comment by akiradeveloper** [comment]
@disaster123  Great.

So I will merge the two branches into master and release v2.0.5.

**Comment by disaster123** [comment]
Thanks!

**Comment by onlyjob** [comment]
Now when initialisation is fast, deconstruction is slower so there are benefits of parallelising it. I'm thinking about situation when UPS is shutting down server because battery have only ~10% charge left so we want shut down (including flushing caches from _dm-wb_ devices) to finish ASAP. Remember that caching device can be a RAM disk.

Also parallel deconstruction seems to be much safer as there are no parallel allocations(?) and it behaves very well in my testing.

Thanks.

**Comment by disaster123** [comment]
To me parallel initialisation and deconstruction works abolutely fine with the upcoming v2.0.5 release. main problem was the usage of kmalloc for large arrays.

**Comment by akiradeveloper** [comment]
As a view of maintainer of kernel part, I want to keep tools as simple as possible. If the parallelization causes some issue, it must be seen outside dm-writeboost code but will be first reported to me.

I am busy with my own work (not meaning dm-writeboost work, but my freelance work), I don't think I can help them all.

I really prefer tools enter and leave dm-writeboost kindly.

If you want to parallelize initialization/deconstruction no matter what. Please show me the code-level evidence that calling `dmsetup` commands are reentrant. Asking dm maintainers is ok. I am really skeptical about this.

**Comment by disaster123** [comment]
Don't know if you mean @onlyjob or me. It works absolutely fine without any change. It was just the memory preassure which was hitted easier with parallelisation. So absolutely nothing todo.

**Comment by akiradeveloper** [comment]
@disaster123 I am neutral if I am talking about technicals.

Parallel initialization sounds very dangerous to me. To begin with, I am skeptical if it's really parallelized. dmsetup might serialize input commands internally.

**Comment by marcin-github** [comment]
Hi!
I seen similar problem to described here (kernel 3.19.8-gentoo, dm-writeboost-2.0.5) but I can't reproduce it. When I run `writeboost` again it finished without problem. I used dm-writeback-tools to set it up. Due to I don't have SDD I used created both devices on LV device. Should I open new issue, hold till this issue will be fixed&closed or help here in some way?
"hdd" has 50GB and pseudo "SSD" has 5GB capacity.

[16942.010400] device-mapper: writeboost: Superblock Header: Magic number invalid
[16960.266086] dmsetup: page allocation failure: order:8, mode:0x2040d0
[16960.266093] CPU: 2 PID: 3520 Comm: dmsetup Tainted: P           O   3.19.8-gentoo #2
[16960.266095] Hardware name:                  /D975XBX, BIOS BX97510J.86A.1487.2007.0902.1724 09/02/2007
[16960.266097]  0000000000000000 000000007b496f05 ffff880001bab768 ffffffff814f24cf
[16960.266101]  0000000000000000 00000000002040d0 ffff880001bab7f8 ffffffff8110a5f6
[16960.266103]  ffffffff8187be80 0000000000000000 ffff880001bab7f8 ffffffff8110d4be
[16960.266106] Call Trace:
[16960.266114]  [<ffffffff814f24cf>] dump_stack+0x45/0x57
[16960.266118]  [<ffffffff8110a5f6>] warn_alloc_failed+0xf6/0x150
[16960.266121]  [<ffffffff8110d4be>] ? __alloc_pages_direct_compact+0x13e/0x140
[16960.266124]  [<ffffffff8110df8e>] __alloc_pages_nodemask+0xace/0xbd0
[16960.266128]  [<ffffffff811531aa>] new_slab+0x2fa/0x450
[16960.266131]  [<ffffffff811536c1>] __slab_alloc.constprop.57+0x3c1/0x4d0
[16960.266134]  [<ffffffff811549d5>] ? kmem_cache_alloc+0x1b5/0x230
[16960.266138]  [<ffffffffc0eec519>] ? resume_cache+0x1d9/0x1310 [dm_writeboost]
[16960.266142]  [<ffffffff8112668e>] ? do_kmem_cache_create+0xce/0x170
[16960.266145]  [<ffffffff811549ae>] kmem_cache_alloc+0x18e/0x230
[16960.266148]  [<ffffffffc0eec4bc>] ? resume_cache+0x17c/0x1310 [dm_writeboost]
[16960.266150]  [<ffffffffc0eec519>] ? resume_cache+0x1d9/0x1310 [dm_writeboost]
[16960.266153]  [<ffffffffc0eec519>] resume_cache+0x1d9/0x1310 [dm_writeboost]
[16960.266167]  [<ffffffffc020659c>] ? dm_get_table_device+0x7c/0x1b0 [dm_mod]
[16960.266170]  [<ffffffffc0eebbb0>] ? prepare_overwrite+0x730/0x940 [dm_writeboost]
[16960.266173]  [<ffffffffc0eea116>] 0xffffffffc0eea116
[16960.266178]  [<ffffffffc020a24b>] dm_table_add_target+0x13b/0x390 [dm_mod]
[16960.266183]  [<ffffffffc020d664>] table_load+0x144/0x360 [dm_mod]
[16960.266187]  [<ffffffffc020d520>] ? retrieve_status+0x1c0/0x1c0 [dm_mod]
[16960.266192]  [<ffffffffc020e273>] ctl_ioctl+0x233/0x520 [dm_mod]
[16960.266197]  [<ffffffffc020e56e>] dm_ctl_ioctl+0xe/0x20 [dm_mod]
[16960.266200]  [<ffffffff81179388>] do_vfs_ioctl+0x2f8/0x500
[16960.266203]  [<ffffffff81179611>] SyS_ioctl+0x81/0xa0
[16960.266206]  [<ffffffff81003f23>] ? do_device_not_available+0x23/0x30
[16960.266209]  [<ffffffff814f7f72>] system_call_fastpath+0x12/0x17
[16960.266211] Mem-Info:
[16960.266212] DMA per-cpu:
[16960.266214] CPU    0: hi:    0, btch:   1 usd:   0
[16960.266216] CPU    1: hi:    0, btch:   1 usd:   0
[16960.266218] CPU    2: hi:    0, btch:   1 usd:   0
[16960.266219] CPU    3: hi:    0, btch:   1 usd:   0
[16960.266220] DMA32 per-cpu:
[16960.266222] CPU    0: hi:  186, btch:  31 usd:   0
[16960.266224] CPU    1: hi:  186, btch:  31 usd:   0
[16960.266225] CPU    2: hi:  186, btch:  31 usd:   0
[16960.266227] CPU    3: hi:  186, btch:  31 usd:   0
[16960.266228] Normal per-cpu:
[16960.266230] CPU    0: hi:  186, btch:  31 usd:   0
[16960.266231] CPU    1: hi:  186, btch:  31 usd:   0
[16960.266233] CPU    2: hi:  186, btch:  31 usd:   0
[16960.266234] CPU    3: hi:  186, btch:  31 usd:   0
[16960.266238] active_anon:309810 inactive_anon:181535 isolated_anon:30
                active_file:251979 inactive_file:245892 isolated_file:0
                unevictable:1737 dirty:215 writeback:32 unstable:0
                free:13370 slab_reclaimable:98500 slab_unreclaimable:9098
                mapped:47733 shmem:6479 pagetables:5197 bounce:0
                free_cma:0
[16960.266246] DMA free:15896kB min:28kB low:32kB high:40kB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB isolated(anon):0kB isolated(file):0kB present:15984kB managed:15896kB mlocked:0kB dirty:0kB writeback:0kB mapped:0kB shmem:0kB slab_reclaimable:0kB slab_unreclaimable:0kB kernel_stack:0kB pagetables:0kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:0 all_unreclaimable? yes
[16960.266247] lowmem_reserve[]: 0 3235 4430 4430
[16960.266255] DMA32 free:26316kB min:6208kB low:7760kB high:9312kB active_anon:970040kB inactive_anon:456620kB active_file:747020kB inactive_file:725480kB unevictable:4788kB isolated(anon):0kB isolated(file):0kB present:3388968kB managed:3315212kB mlocked:4788kB dirty:772kB writeback:128kB mapped:137572kB shmem:18084kB slab_reclaimable:286744kB slab_unreclaimable:23308kB kernel_stack:4928kB pagetables:15028kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:0 all_unreclaimable? no
[16960.266256] lowmem_reserve[]: 0 0 1194 1194
[16960.266263] Normal free:11268kB min:2292kB low:2864kB high:3436kB active_anon:269200kB inactive_anon:269520kB active_file:260896kB inactive_file:258088kB unevictable:2160kB isolated(anon):120kB isolated(file):0kB present:1310720kB managed:1223636kB mlocked:2160kB dirty:88kB writeback:0kB mapped:53360kB shmem:7832kB slab_reclaimable:107256kB slab_unreclaimable:13084kB kernel_stack:2320kB pagetables:5760kB unstable:0kB bounce:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:0 all_unreclaimable? no
[16960.266265] lowmem_reserve[]: 0 0 0 0
[16960.266267] DMA: 0*4kB 1*8kB (U) 1*16kB (U) 0*32kB 2*64kB (U) 1*128kB (U) 1*256kB (U) 0*512kB 1*1024kB (U) 1*2048kB (R) 3*4096kB (M) = 15896kB
[16960.266280] DMA32: 5200*4kB (UEM) 112*8kB (UEM) 31*16kB (UMR) 27*32kB (UEMR) 15*64kB (UMR) 12*128kB (UMR) 0*256kB 2*512kB (R) 0*1024kB 0*2048kB 0*4096kB = 26576kB
[16960.266291] Normal: 668*4kB (UEMR) 101*8kB (UEMR) 54*16kB (UEMR) 51*32kB (UEMR) 27*64kB (UEMR) 12*128kB (UEMR) 3*256kB (EMR) 1*512kB (M) 1*1024kB (M) 0*2048kB 0*4096kB = 11544kB
[16960.266304] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[16960.266305] 506746 total pagecache pages
[16960.266307] 894 pages in swap cache
[16960.266309] Swap cache stats: add 22862, delete 21968, find 1170067/1169804
[16960.266311] Free swap  = 6207032kB
[16960.266312] Total swap = 6291452kB
[16960.266313] 1178918 pages RAM
[16960.266315] 0 pages HighMem/MovableOnly
[16960.266316] 36136 pages reserved
[16960.266317] 4096 pages cma reserved
[16960.266319] SLUB: Unable to allocate memory on node -1 (gfp=0xd0)
[16960.266321]   cache: dmwb_rambuf, object size: 524288, buffer size: 1048576, default order: 8, min order: 8
[16960.266323]   dmwb_rambuf debugging increased min order, use slub_debug=O to disable.
[16960.266325]   node 0: slabs: 0, objs: 0, free: 0
[16960.266327] device-mapper: writeboost: Failed to allocate rambuf->data
[16960.266378] device-mapper: writeboost: init_rambuf_pool failed
[16960.266586] device-mapper: table: 253:16: writeboost: resume_cache failed
[16960.266589] device-mapper: ioctl: error adding target to table



meminfo (after a few next tries) is:
# cat /proc/meminfo 
MemTotal:        4554744 kB
MemFree:         2155620 kB
MemAvailable:    2851096 kB
Buffers:           53436 kB
Cached:           646920 kB
SwapCached:         4464 kB
Active:          1098864 kB
Inactive:         991728 kB
Active(anon):     633296 kB
Inactive(anon):   792560 kB
Active(file):     465568 kB
Inactive(file):   199168 kB
Unevictable:       17544 kB
Mlocked:           17544 kB
SwapTotal:       6291452 kB
SwapFree:        6178712 kB
Dirty:               436 kB
Writeback:             0 kB
AnonPages:       1404768 kB
Mapped:           158144 kB
Shmem:             26212 kB
Slab:             117740 kB
SReclaimable:      62708 kB
SUnreclaim:        55032 kB
KernelStack:        7408 kB
PageTables:        20920 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     8568824 kB
Committed_AS:    4997500 kB
VmallocTotal:   34359738367 kB
VmallocUsed:      202892 kB
VmallocChunk:   34359461156 kB
CmaTotal:          16384 kB
CmaFree:            4176 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
DirectMap4k:      138792 kB
DirectMap2M:     4577280 kB


What dm-writeback do if found garbage in cache device, maybe some kind of unexpected magic bytes could made problem? Just guessing.

**Comment by akiradeveloper** [comment]
Hi Marcin,

> Should I open new issue, hold till this issue will be fixed&closed or help here in some way?

No.

I think you failed to allocate ram buffer.

```
[16960.266321] cache: dmwb_rambuf, object size: 524288, buffer size: 1048576, default order: 8, min order: 8
[16960.266323] dmwb_rambuf debugging increased min order, use slub_debug=O to disable.
[16960.266325] node 0: slabs: 0, objs: 0, free: 0
[16960.266327] device-mapper: writeboost: Failed to allocate rambuf->data
[16960.266378] device-mapper: writeboost: init_rambuf_pool failed
```

Now it's becoming clear that using kmalloc for rambuf and writeboost_segs isn't proper way of allocating these big memory chunks.

**Comment by disaster123** [comment]
But it's still the fastest memory available as it is preallocated. I think dm-writeboost should still try to use kmalloc and only fallback.

Like here:
https://www.redhat.com/archives/cluster-devel/2013-June/msg00029.html
or
http://www.gossamer-threads.com/lists/linux/kernel/2174163
or
http://www.gossamer-threads.com/lists/linux/kernel/1945294
or
http://www.gossamer-threads.com/lists/linux/kernel/1932033
or
http://www.gossamer-threads.com/lists/linux/kernel/1723865
or
https://groups.google.com/forum/#!topic/linux.kernel/X0hjybwTljM
;-)

**Comment by akiradeveloper** [comment]
@disaster123 The implementation won't be as easy as you imagine. The code will be very dirty. I think we first try with vmalloc to make sure bugs disappear and leave the fallback plan for the future enhancement.

**Comment by disaster123** [comment]
OK you're right. Stability is more important than performance.

**Comment by akiradeveloper** [comment]
I am working on #77

**Comment by akiradeveloper** [comment]
I've pushed fix-buf-alloc branch as a candidate for v2.0.6.

Please retest with this branch. It works fine with me.

**Comment by marcin-github** [comment]
I try it, I didn't had any issue. But  I couldn't reproduce it even second time with v2.0.5.

**Comment by akiradeveloper** [comment]
Thanks you Marcin.

I think your case is quite hard to reproduce. You tried to allocate eight 512KB contiguous buffers and failed. It's possible but close to miracle.

**Comment by marcin-github** [comment]
Today I got the same problem. Fresh booted OS, when I created wb_device then:
[  415.561770] device-mapper: writeboost: Superblock Header: Magic number invalid
[  441.152312] dmsetup: page allocation failure: order:5, mode:0x40d0
[  441.152320] CPU: 0 PID: 2349 Comm: dmsetup Tainted: G           O    4.3.3-aufs #1
[  441.152322] Hardware name: Gigabyte Technology Co., Ltd. 965P-DS3/965P-DS3, BIOS F14 06/25/2009
[  441.152325]  0000000000000000 0000000018e8489e ffff8800335739f0 ffffffff8128902f
[  441.152330]  00000000000040d0 ffff880033573a80 ffffffff81103544 ffffffe0000040d0
[  441.152333]  0000000000000246 0000002000000005 0000000000000005 0000000000000040
[  441.152337] Call Trace:
[  441.152348]  [<ffffffff8128902f>] dump_stack+0x44/0x55
[  441.152353]  [<ffffffff81103544>] warn_alloc_failed+0xf4/0x150
[  441.152357]  [<ffffffff81106be7>] __alloc_pages_nodemask+0x6e7/0x890
[  441.152360]  [<ffffffff81107022>] alloc_kmem_pages+0x12/0x20
[  441.152364]  [<ffffffff8111dff3>] kmalloc_order+0x13/0x70
[  441.152367]  [<ffffffff8111e06f>] kmalloc_order_trace+0x1f/0x90
[  441.152372]  [<ffffffffc01ee46e>] 0xffffffffc01ee46e
[  441.152386]  [<ffffffffc00059ab>] dm_table_add_target+0x13b/0x380 [dm_mod]
[  441.152393]  [<ffffffffc0008caf>] table_load+0x12f/0x340 [dm_mod]
[  441.152400]  [<ffffffffc0008b80>] ? retrieve_status+0x1b0/0x1b0 [dm_mod]
[  441.152407]  [<ffffffffc000989a>] ctl_ioctl+0x22a/0x4e0 [dm_mod]
[  441.152415]  [<ffffffffc0009b5e>] dm_ctl_ioctl+0xe/0x20 [dm_mod]
[  441.152418]  [<ffffffff8116303f>] do_vfs_ioctl+0x2bf/0x490
[  441.152421]  [<ffffffff8116324c>] SyS_ioctl+0x3c/0x70
[  441.152425]  [<ffffffff81474b1b>] entry_SYSCALL_64_fastpath+0x16/0x6e
[  441.152427] Mem-Info:
[  441.152433] active_anon:20353 inactive_anon:12565 isolated_anon:0
                active_file:50049 inactive_file:49807 isolated_file:0
                unevictable:3289 dirty:0 writeback:0 unstable:0
                slab_reclaimable:69874 slab_unreclaimable:5027
                mapped:5931 shmem:27 pagetables:604 bounce:0
                free:6746 free_pcp:0 free_cma:3945
[  441.152443] DMA free:3972kB min:60kB low:72kB high:88kB active_anon:416kB inactive_anon:516kB active_file:3008kB inactive_file:2000kB unevictable:204kB isolated(anon):0kB isolated(file):0kB present:15984kB managed:15900kB mlocked:204kB dirty:0kB writeback:0kB mapped:436kB shmem:0kB slab_reclaimable:4480kB slab_unreclaimable:428kB kernel_stack:112kB pagetables:60kB unstable:0kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB writeback_tmp:0kB pages_scanned:84 all_unreclaimable? no
[  441.152446] lowmem_reserve[]: 0 976 976 976
[  441.152456] DMA32 free:23012kB min:3964kB low:4952kB high:5944kB active_anon:80996kB inactive_anon:49744kB active_file:197188kB inactive_file:197228kB unevictable:12952kB isolated(anon):0kB isolated(file):0kB present:1031040kB managed:1004256kB mlocked:12952kB dirty:0kB writeback:0kB mapped:23288kB shmem:108kB slab_reclaimable:275016kB slab_unreclaimable:19680kB kernel_stack:2080kB pagetables:2356kB unstable:0kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:15780kB writeback_tmp:0kB pages_scanned:0 all_unreclaimable? no
[  441.152459] lowmem_reserve[]: 0 0 0 0
[  441.152463] DMA: 61*4kB (EM) 20*8kB (EM) 9*16kB (EM) 3*32kB (EM) 4*64kB (EM) 0*128kB 2*256kB (UM) 1*512kB (E) 2*1024kB (EM) 0*2048kB 0*4096kB = 3972kB
[  441.152480] DMA32: 627*4kB (UEMC) 190*8kB (UM) 89*16kB (UMC) 40*32kB (UMC) 11*64kB (UM) 0*128kB 1*256kB (C) 0*512kB 1*1024kB (C) 1*2048kB (C) 3*4096kB (C) = 23052kB
[  441.152497] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[  441.152499] 101034 total pagecache pages
[  441.152503] 138 pages in swap cache
[  441.152505] Swap cache stats: add 229, delete 91, find 0/0
[  441.152507] Free swap  = 3144808kB
[  441.152509] Total swap = 3145724kB
[  441.152510] 261756 pages RAM
[  441.152512] 0 pages HighMem/MovableOnly
[  441.152513] 6717 pages reserved
[  441.152514] 4096 pages cma reserved
[  441.159578] device-mapper: table: 254:2: writeboost: init_read_cache_cells failed
[  441.159585] device-mapper: ioctl: error adding target to table

After this I did backup of caching device (for the case when another attempt creating wb_device will finish with such error), and tried to create wb_device. Second time dm-writeboost-tools created device without problem.
Maybe I'll add to scripts two lines with dd wich will do backup caching device before trying creating wb_device.

**Comment by akiradeveloper** [comment]
it failed to allocate read_cache_cells.

How big is your cache device?
How much memory does your machine have? (261756 pages?)

**Comment by marcin-github** [comment]
cache device is 10GB, RAM=1GB

**Comment by akiradeveloper** [comment]
@marcin-github Thanks for reporting. Is this reproducible?

**Comment by akiradeveloper** [comment]
```c
	for (i = 0; i < cells->size; i++) {
		struct read_cache_cell *cell = cells->array + i;
		cell->data = kmalloc(1 << 12, GFP_KERNEL);
```

this could be fixed by using vmalloc for `cell->data`

**Comment by marcin-github** [comment]
No, it's not reproducible.

**Comment by akiradeveloper** [comment]
I will fix this issue by using vmalloc for the cell buffers. But I don't know it works

**Comment by akiradeveloper** [comment]
v2.1.2 uses vmalloc for cell buffers.

**Comment by akiradeveloper** [comment]
I'm going to close this issue.

@marcin-github If you find another allocation problem please open the other issue.
